New ASCII APIs

adamsitnik opened the pull request:
fixes #28230

```C#
namespace System.Text;

public static class Ascii
{   
    public static bool IsValid(ReadOnlySpan<byte> value);
    public static bool IsValid(ReadOnlySpan<char> value);

    public static bool IsValid(byte value);
    public static bool IsValid(char value);

    public static OperationStatus ToUpper(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten);
    public static OperationStatus ToUpper(ReadOnlySpan<char> source, Span<char> destination, out int charsWritten);
    public static OperationStatus ToUpper(ReadOnlySpan<byte> source, Span<char> destination, out int charsWritten);
    public static OperationStatus ToUpper(ReadOnlySpan<char> source, Span<byte> destination, out int bytesWritten);

    public static OperationStatus ToLower(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten);
    public static OperationStatus ToLower(ReadOnlySpan<char> source, Span<char> destination, out int charsWritten);
    public static OperationStatus ToLower(ReadOnlySpan<byte> source, Span<char> destination, out int charsWritten);
    public static OperationStatus ToLower(ReadOnlySpan<char> source, Span<byte> destination, out int bytesWritten);

    public static OperationStatus ToUpperInPlace(Span<byte> value, out int bytesWritten);
    public static OperationStatus ToUpperInPlace(Span<char> value, out int charsWritten);

    public static OperationStatus ToLowerInPlace(Span<byte> value, out int bytesWritten);
    public static OperationStatus ToLowerInPlace(Span<char> value, out int charsWritten);

    public static OperationStatus ToUtf16(ReadOnlySpan<byte> source, Span<char> destination, out int charsWritten);
    public static OperationStatus FromUtf16(ReadOnlySpan<char> source, Span<byte> destination, out int bytesWritten);

    public static Range Trim(ReadOnlySpan<byte> value);
    public static Range Trim(ReadOnlySpan<char> value);
    public static Range TrimStart(ReadOnlySpan<byte> value);
    public static Range TrimStart(ReadOnlySpan<char> value);
    public static Range TrimEnd(ReadOnlySpan<byte> value);
    public static Range TrimEnd(ReadOnlySpan<char> value);
}
```

GrabYourPitchforks committed:
Initial ASCII methods

GrabYourPitchforks committed:
Add transcoding APIs

GrabYourPitchforks committed:
Implement Trim

GrabYourPitchforks committed:
Split ASCII utilities into separate files

GrabYourPitchforks committed:
Add ref asm

GrabYourPitchforks committed:
Fun with case conversion!

GrabYourPitchforks committed:
Fun with case conversion!

GrabYourPitchforks committed:
Updates!

GrabYourPitchforks committed:
Fix incorrect comparison

GrabYourPitchforks committed:
Fix incorrect precondition checks

GrabYourPitchforks committed:
Update main vectorized loop

GrabYourPitchforks committed:
Perf improvements & fix arithmetic error

adamsitnik committed:
tests for Ascii.GetIndexOfFirstNonAsciiByte

adamsitnik committed:
tests for Ascii.GetIndexOfFirstNonAsciiChar

adamsitnik committed:
add tests for Ascii.IsAscii

adamsitnik committed:
add tests for Ascii.FromUtf16

adamsitnik committed:
add tests for Ascii.ToUtf16

adamsitnik committed:
add tests for Ascii.Trim* and fix bug they have discovered

adamsitnik committed:
ToUpper & ToLower tests

adamsitnik committed:
implement the missing pieces for case conversions + fix the tests

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into asciiAPIs

adamsitnik committed:
implement TryToLowerInPlace/TryToUpperInPlace

adamsitnik committed:
implement Ascii.StartsWith* and EndsWith* methods

adamsitnik committed:
implement Ascii.Equals* methods

adamsitnik committed:
use self-describing names at a cost of using pragma disable ;)

adamsitnik committed:
throw ArgumentException with meaningful error message

adamsitnik committed:
rename files

adamsitnik committed:
Implement IndexOf and LastIndexOf using narrowing and widening

adamsitnik committed:
Implement IndexOfIgnoreCase and LastIndexOfIgnoreCase

adamsitnik commented:
@GrabYourPitchforks the things I would like to discuss:

1. For the conversion APIs the number of elements consumed is always the same as number of elements written. 

https://github.com/dotnet/runtime/blob/2d9105e5816205d5420e8de6e6ca7bfe33f5a961/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.CaseConversion.cs#L95-L96

IMO we could simplify the API and have only one parameter:

```diff
- OperationStatus ToUpper(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten);
+ OperationStatus ToUpper(ReadOnlySpan<byte> source, Span<byte> destination, out int elementsProcessed);
```

2. There are few new methods that operate on two buffers of characters that try to solve a problem which was already solved by existing methods. The only difference is that they throw when "needle" contains non-ASCII char. 

Example:

```cs
// proposed
Ascii.StartsWithIgnoreCase(ReadOnlySpan<char> text, ReadOnlySpan<char> value)
// users can already use:
MemoryExtensions.StartsWith(ReadOnlySpan<char> text, ReadOnlySpan<char> value, StringComparison.OrdinalIgnoreCase)
```

Since checking if "needle" contains non-ASCII chars costs some extra CPU cycles and users should typically have a good control over the needle I wonder if we should introduce such methods.

adamsitnik committed:
refactoring

adamsitnik committed:
IsAscii methods

adamsitnik committed:
*GetHashCode(chars)

adamsitnik committed:
*GetHashCode(bytes)

adamsitnik committed:
solve buffer overrun: 8 chars need to be narrowed to 8 (not 16 like before) bytes in case of ASCII case conversion

adamsitnik committed:
disable the tests that are failing due to Mono bug

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into asciiAPIs

stephentoub commented:
Thanks for working on this, @adamsitnik.

There are a bunch of places these methods could be used, e.g. just a few examples:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs#L13
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs#L42
https://github.com/dotnet/runtime/blob/20dd6ea921028b3a2cc660654b65112077671fb5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs#L980
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs#L143-L148
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/System.Private.Uri/src/System/UriHelper.cs#L136-L144
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/System.Private.Uri/src/System/UriHelper.cs#L173-L180
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/Common/src/System/Net/CaseInsensitiveAscii.cs#L8
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Http/src/System/Net/Http/Headers/UriHeaderParser.cs#L60
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs#L156-L168
https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs#L500
https://github.com/dotnet/runtime/blob/c59b5171e4fb2b000108bec965f8ce443cb95a12/src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs#L491
... etc.
plus we should replace all existing direct usage of ASCIIUtility with these public API whereever it makes sense to do so.

There are also some related uses that can't be replaced as-is but that suggest we could use some additional IndexOf* helpers, e.g. IndexOfRange/IndexOfAnyExceptRange:
https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/Common/src/System/Net/HttpValidationHelpers.cs#L24-L34

danmoseley commented:
In that last example ContainsNonAsciiChars does !IsAscii(ROS<char>) not suffice?

stephentoub commented:
> In that last example ContainsNonAsciiChars does !IsAscii(ROS<char>) not suffice?

No, because despite the name, it's checking for both > 0x7e (not 0x7f) and < 0x20.

danmoseley commented:
Oh - yes. Bad name.

adamsitnik commented:
Also, based on how frequently these methods are going to be used I can optimize them further

adamsitnik committed:
fix a bug (tests that are not compiled are always passing)

adamsitnik committed:
use new APIs across BCL:

use Trim and EqualsIgnoreCase in CharArrayHelpers (used by System.Net.Http and System.Net.Http.WinHttpHandler)
use Equals in various System.Net.Http places
use EqualsIgnoreCase in CaseInsensitiveAscii (used by System.Net.WebHeaderCollection), force it's inlining
use TryToLowerInPlace in System.Net.HttpListener
use ToLower in QPackEncoder (used by System.Net.Http)
use IndexOf in Http2LoopbackConnection (used by System.Net.Http* test projects)
use IsAscii and FromUtf16 in PInvokeMarshal (NatvieAOT)

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.CaseConversion.cs:
```diff
@@ -0,0 +1,507 @@
+            for (; i < elementCount; i++)
+            {
+                uint element = uint.CreateTruncating(pSrc[i]);
+                if (!UnicodeUtility.IsAsciiCodePoint(element)) { break; }
```
Nit: let's avoid single-line if blocks with bodies like this, especially in the context of all the other blocks being multi-line

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.CaseConversion.cs:
```diff
@@ -0,0 +1,507 @@
+            }
+            else if (Vector64.IsHardwareAccelerated)
+            {
+                narrow.GetLower().StoreUnsafe(ref *pDest, destOffset);
```
Does the Sse2 path produce better code than this path, or why is it needed?

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Comparison.cs:
```diff
@@ -0,0 +1,412 @@
+using SkipChecks = System.Boolean;
+using CheckBytes = System.Byte;
+using CheckChars = System.Char;
+using CheckNonAscii = System.Byte;
```
I don't understand the purpose of this. Especially with multiple names mapped to the same type.

adamsitnik replied:
> I don't understand the purpose of this.

The private `Equals` method compares `ROS<char>` with `ROS<byte>`:

```cs
Equals<TCheck>(ReadOnlySpan<char> chars, ReadOnlySpan<byte> bytes) where TCheck : struct
```

is used by the following public methods:

```cs
public static bool Equals(ReadOnlySpan<byte> left, ReadOnlySpan<char> right)
public static unsafe bool StartsWith(ReadOnlySpan<byte> text, ReadOnlySpan<char> value)
public static unsafe bool EndsWith(ReadOnlySpan<byte> text, ReadOnlySpan<char> value)
public static unsafe bool StartsWith(ReadOnlySpan<char> text, ReadOnlySpan<byte> value)
public static unsafe bool EndsWith(ReadOnlySpan<char> text, ReadOnlySpan<byte> value)
```

`Equals` should not throw when any of the provided arguments contain non-ascii chars.
But `StartsWith` and `EndsWith` should throw only when `value` contains non-ascii chars. The thing is that value can be both `char` and `byte`. My goal of using the aliases was not express "in this case bytes are the value, so check the bytes" (same for chars) and "don't perform any checks". 

> Especially with multiple names mapped to the same type.

That is true. I will try to come up with simpler solution.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Comparison.cs:
```diff
@@ -0,0 +1,412 @@
+{
+    public static partial class Ascii
+    {
+        public static bool TryGetHashCode(ReadOnlySpan<byte> value, out int hashCode)
```
What's the use case for these case-sensitive APIs?  Why would you want to hash bytes but only if they're ASCII, especially when you can just call IsAscii yourself?

You can hash any `ReadOnlySpan<byte>` with:
```C#
HashCode hc = default;
hc.AddBytes(value);
... = hc.ToHashCode();
```
and if it was really valuable to be able to do that with less ceremony, we could add one-shot static methods to let you do:
```C#
... = HashCode.Combine(spanOfBytes);
```
at which point you can do the ASCII check yourself, e.g.
```C#
if (Ascii.IsAscii(value)) { Use(HashCode.Combine(spanOfBytes)); }
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Comparison.cs:
```diff
@@ -0,0 +1,412 @@
+                }
+                while (!Unsafe.IsAddressGreaterThan(ref currentCharsSearchSpace, ref oneVectorAwayFromCharsEnd));
+
+                // If any elements remain, process the first vector in the search space.
```
> process the first vector in the search space

The last vector?

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Searching.cs:
```diff
@@ -0,0 +1,278 @@
+        }
+    }
+}
+#pragma warning restore SA1121 // Use built-in type alias
```
```suggestion
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Transcoding.cs:
```diff
@@ -0,0 +1,86 @@
+            fixed (byte* pSource = &MemoryMarshal.GetReference(source))
+            fixed (char* pDestination = &MemoryMarshal.GetReference(destination))
+            {
+                nuint numElementsActuallyConverted = ASCIIUtility.WidenAsciiToUtf16(pSource, pDestination, numElementsToConvert);
```
Do we still need ASCIIUtility, or can the contents of that type be folded into Ascii?

adamsitnik replied:
I would love to fold it and also get rid of all the pinning and use managed references (`ref`) everywhere.

@stephentoub Is there any chance you could review https://github.com/dotnet/runtime/pull/73055? Once I merge it I could fold it and avoid merge conflict.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Trimming.cs:
```diff
@@ -0,0 +1,58 @@
+        private static Range TrimHelper<T>(ReadOnlySpan<T> value, TrimType trimType)
+            where T : unmanaged, IBinaryInteger<T>
+        {
+            const uint trimMask =
```
Nit: trimMask => TrimMask

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:
```diff
@@ -977,8 +977,9 @@ public static bool ParticipatesInCaseConversion(ReadOnlySpan<char> s)
-        public static bool IsAscii(ReadOnlySpan<char> s) // TODO https://github.com/dotnet/runtime/issues/28230: Replace once Ascii is available
+        public static bool IsAscii(ReadOnlySpan<char> s)
         {
+#if NETSTANDARD
```
Elsewhere uses:
```C#
#if REGEXGENERATOR
```

adamsitnik commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackEncoder.cs:
```diff
@@ -2,7 +2,8 @@
 
 #nullable enable
-using System.Collections.Generic;
+using System.Buffers;
```
Thanks, I was not aware of that. I wonder which option would be better: 

1. revert these changes for now 
2. wrap them with `#if NET8_0_OR_GREATER` which would do nothing as of today, but I should work OOTB in a few weeks from now.

I am most likely going to just revert them.

stephentoub replied:
Yeah, let's just revert for now.

adamsitnik commented on src/libraries/System.Net.Mail/src/System/Net/Mail/DotAtomReader.cs:
```diff
@@ -43,7 +44,7 @@ internal static bool TryReadReverse(string data, int index, out int outIndex, bo
             for (; 0 <= index; index--)
             {
-                if (data[index] <= MailBnfHelper.Ascii7bitMaxValue // Any Unicode allowed
+                if (Ascii.IsAscii(data[index]) // Any ASCII allowed
```
> I assume char.IsAscii(char) and Ascii.IsAscii(char) are exactly the same?

That is correct.

> Do we need both? Is the latter just there for symmetry with IsAscii(byte)?

@GrabYourPitchforks is the originator of the API, I expect that it's for the symmetry with `IsAscii(byte)` and `IsAscii(ROSC<byte/char>)`

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.CaseConversion.cs:
```diff
@@ -0,0 +1,507 @@
+    public static partial class Ascii
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static OperationStatus ToUpper(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
```
That is still on my TODO list, I am delaying that until the APIs get officially approved.

adamsitnik committed:
Apply suggestions from code review

Co-authored-by: Stephen Toub <stoub@microsoft.com>

adamsitnik committed:
address code review feedback

adamsitnik commented on src/libraries/Common/src/System/CharArrayHelpers.cs:
```diff
@@ -37,11 +44,17 @@ internal static bool EqualsOrdinalAsciiIgnoreCase(string left, char[] right, int
             DebugAssertArrayInputs(array, startIndex, length);
 
+#if NET7_0_OR_GREATER
+            Range range = Buffers.Text.Ascii.Trim(new ReadOnlySpan<char>(array, startIndex, length));
```
great catch! I am going to revert it and let Networking folks to use the new APIs and decide on their own (this method is usded by networking stack and I don't know whether in given context ascii chars are allowed or not)

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:
```diff
@@ -1448,12 +1448,10 @@ private Task WriteAsciiStringAsync(string s, bool async)
-                    writeBuffer[offset++] = (byte)c;
-                }
-                _writeOffset = offset;
+                OperationStatus operationStatus = Ascii.FromUtf16(s, _writeBuffer.AsSpan(offset), out _, out int bytesWritten);
```
The loop above in WriteStringAsync:
https://github.com/dotnet/runtime/pull/75012/files#diff-595f700a4c83b735d856fe14b5ebe8c7f7c18ae6e586adc8b398a997a0678274L1391-L1398
can also be replaced.

stephentoub commented:
Another place this could be used:
https://github.com/dotnet/runtime/blob/cc9d46546b8e72b34a5f387a5ec9615197d79f1f/src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs#L214-L222

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into asciiAPIs

# Conflicts:
#	src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

adamsitnik committed:
fold ASCIIUtility into Ascii, use public APIs

adamsitnik committed:
fix tests that were relying on reflection so far (and I did not know about their existence)

adamsitnik committed:
fix byte->char casting

stephentoub commented:
@adamsitnik, what are the next steps here?  Do you need assistance?  Can we separate out the uncontroversial members, get those PR'd and merged, and then focus on the design for the remainder?

davidfowl commented:
cc @BrennanConroy

adamsitnik committed:
adjust code after recent API Review

adamsitnik committed:
add missing XML docs

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into asciiAPIs

# Conflicts:
#	src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs
#	src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs
#	src/libraries/System.Private.Uri/src/System/UriHelper.cs

adamsitnik committed:
cleanup

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Text/Ascii.CaseConversion.cs:
```diff
@@ -0,0 +1,605 @@
+            }
+        }
+
+        private struct ToUpperConversion { }
+        private struct ToLowerConversion { }
```
Could there be used static abstract interfaces instead?

MihaZupan commented on src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:
```diff
@@ -200,10 +195,14 @@ public static string IdnEquivalent(string hostname)
             {
                 // just lowercase for ascii
-                return hostname.ToLowerInvariant();
+                return string.Create(hostname.Length, hostname, static (chars, asciiHostName) =>
+                {
+                    OperationStatus status = Ascii.ToLower(asciiHostName, chars, out _);
+                    Debug.Assert(status == OperationStatus.Done);
+                });
```
```suggestion
                return hostname.ToLowerInvariant();
```

The common case here is that the `ToLowerInvariant` will see the input is already lowercase and return the original instance. We don't want the additional allocation.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Trimming.cs:
```diff
@@ -0,0 +1,80 @@
+            {
+                for (; start < value.Length; start++)
+                {
+                    uint elementValue = uint.CreateTruncating(value[start]);
+                    if ((elementValue > 0x20) || ((TrimMask & (1u << ((int)elementValue - 1))) == 0))
```
You could make this branch-free with:
```C#
        uint c = (ushort)(uint.CreateTruncating(value) - '\t');
        if ((int)((0xF8000100U << (short)c) & (c - 32)) >= 0)
```
See https://github.com/dotnet/runtime/blob/2480f0104160f8218d55568771b57c7f9c0d02f5/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L4599-L4628 for an explanation.

adamsitnik replied:
If possible I would prefer to avoid adding further optimizations now, merge the API, add benchmarks to dotnet/performance and then let others tune it.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Text/Ascii.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <param name="value">The value to inspect.</param>
+        /// <returns>True if <paramref name="value"/> contains only ASCII bytes or is
+        /// empty; False otherwise.</returns>
+        public static unsafe bool IsValid(ReadOnlySpan<byte> value) => value.IsEmpty || GetIndexOfFirstNonAsciiByte(value) < 0;
```
We expect empty to be so common that it's worth special-casing here and doing the check twice when it's not empty?  The very first thing GetIndexOfFirstNonAsciiByte does is the same IsEmpty check.

stephentoub commented on src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:
```diff
@@ -200,10 +195,14 @@ public static string IdnEquivalent(string hostname)
             {
                 // just lowercase for ascii
-                return hostname.ToLowerInvariant();
+                return string.Create(hostname.Length, hostname, static (chars, asciiHostName) =>
```
Why is this code better than what was there before?  We still want to avoid the allocation in the case where the string is already all lowercased.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:
```diff
@@ -990,6 +991,9 @@ public static bool ParticipatesInCaseConversion(ReadOnlySpan<char> s)
 
             return true;
+#else
+            return Text.Ascii.IsValid(s);
```
Can we add `using System.Text;` and then remove the `Text.` from here?

adamsitnik committed:
address code review feedback

adamsitnik commented on src/libraries/System.Net.HttpListener/src/System/Net/HttpListener.cs:
```diff
@@ -179,6 +165,52 @@ internal void AddPrefix(string uriPrefix)
+
+                    if (Ascii.ToLowerInPlace(destination.Slice(0, toLowerLength), out _) != OperationStatus.Done)
+                    {
+                        throw new IndexOutOfRangeException(); // backward compat for non-ASCII characters
```
Good catch, I've missed the fact that `CaseInsensitiveAscii.AsciiToLower` defines all 255 values, rather than just 0-127.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Text/ASCIIEncoding.cs:
```diff
@@ -188,11 +189,12 @@ private protected sealed override unsafe int GetByteCountFast(char* pChars, int
             if (!(fallback is EncoderReplacementFallback replacementFallback
                 && replacementFallback.MaxCharCount == 1
-                && replacementFallback.DefaultString[0] <= 0x7F))
+                && Ascii.IsValid(replacementFallback.DefaultString[0])))
```
`Ascii.IsValid(char)` was added mostly for completeness and having everything in once place.

> Should we have Array.IsValid(int value)

I assume you meant `Ascii.IsValid(int)`? On the one hand it would be more flexible (users could pass integers as inputs too), on the other I am not sure about codegen differences and whether everyone knows that they can pass byte/char to int-accepting method.

stephentoub replied:
> Ascii.IsValid(char) was added mostly for completeness and having everything in once place.

So which should we be using?  Should we change all use of char.IsAscii to be Ascii.IsValid?

MihaZupan replied:
Please no :) Personally I find `char.IsAscii` to be more readable.

stephentoub replied:
Then should we replace all Ascii.IsValid(char) with char.IsAscii?

We "just" added char.IsAscii in .NET 6.  Now in .NET 8 we're adding an identical Ascii.IsValid(char).  I'm trying to rationalize why we have both.  I get the consistency argument, but I don't think it's worth consistency just to have a method no one uses.

stephentoub replied:
@adamsitnik, opinions?

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Text/Ascii.CaseConversion.cs:
```diff
@@ -0,0 +1,605 @@
+            where TTo : unmanaged, IBinaryInteger<TTo>
+            where TCasing : struct
+        {
+            if (MemoryMarshal.AsBytes(source).Overlaps(MemoryMarshal.AsBytes(destination)))
```
`TFROM` and `TTO` can be different (char and byte for example) and in theory someone could do sth like this:

https://github.com/dotnet/runtime/blob/c0f38d1689e4adedd661385ba8fe7c931fad3ba6/src/libraries/System.Text.Encoding/tests/Ascii/CaseConversionTests.cs#L28-L29

Since everything can be casted to bytes I decided to use it and cover all possible scenarios, but to be honest I did it mostly to cover all possible unit testing scenarios rather than expecting someone to actually do it.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into asciiAPIs

adamsitnik committed:
Update src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>

jeffhandley reviewed (approved):
I reviewed this mostly through the lens of how the new methods are _used_--each usage is a nice improvement in maintainability. I also appreciate the copious comments through these implementations and tests.

dakersnar commented:
More regressions: https://github.com/dotnet/perf-autofiling-issues/issues/11194

lewing commented:
mono interpreter regression https://github.com/dotnet/perf-autofiling-issues/issues/11147
there is likely a wasm regression as well but it won't show up until the benchmarks are fixed there

EgorBo commented:
It seems this PR still has regressions opened, e.g. these 5 (we don't have a lot of coverage for ToLower so only 5): https://github.com/dotnet/perf-autofiling-issues/issues/10226 - presumably, the new logic is a bit slower than from https://github.com/dotnet/runtime/pull/78262 it removed

adamsitnik commented:
@EgorBo we have an issue that is tracking it: https://github.com/dotnet/runtime/issues/80245

> My guess that it's because the SIMD part to change case was a bit more efficient in

The new methods support 4 different combinations of input and output: byte->byte, byte->char, char->char and char->byte. The generic code is most likely not as optimal as it was for char->char implementation we had:

https://github.com/dotnet/runtime/blob/1c442fcb36e73eb3d8be10423a45389ce2a468b9/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.CaseConversion.cs#L250-L260

+ there might be some overhead for pinning the input twice (https://github.com/dotnet/runtime/pull/80709)
